GoogleApiWrappers/src/Google.Apis.Storage.v1.ClientWrapper/StorageClient.DownloadObject.cs

```diff
@@ -0,0 +1,165 @@
+            CancellationToken cancellationToken,
+            IProgress<IDownloadProgress> progress)
+        {
+            string uri = GetUri(source);
+            return DownloadObjectAsyncImpl(uri, destination, options, cancellationToken, progress);
+        }
+
+        /// <summary>
+        /// Constructs the media URL of an object from its bucket and name.
+        /// </summary>
+        private static string GetUri(string bucket, string sourceName)
+        {
+            bucket.CheckNotNull(nameof(bucket));
+            sourceName.CheckNotNull(nameof(sourceName));
+            return $"https://www.googleapis.com/download/storage/v1/b/{Uri.EscapeDataString(bucket)}/o/{Uri.EscapeDataString(sourceName)}?alt=media";
+        }
+
+        /// <summary>
+        /// Obtains the media URL of an object from eiteher the MediaLink property, or its bucket and name.
+        /// </summary>
+        private static string GetUri(Object source)
+        {
+            source.CheckNotNull(nameof(source));
+            if (source.MediaLink != null)
+            {
+                return source.MediaLink;
+            }
+            if (source.Bucket != null && source.Name != null)
+            {
+                return GetUri(source.Bucket, source.Name);
+            }
+            throw new ArgumentException("Object must have either MediaLink or both Bucket and Name set");
+        }
+
+        private void DownloadObjectImpl(string uri, Stream destination, DownloadObjectOptions options, IProgress<IDownloadProgress> progress)
+        {
+            // URI will definitely not be null; that's constructed internally.
+            destination.CheckNotNull(nameof(destination));
+            var downloader = new MediaDownloader(Service);
+            options?.ModifyDownloader(downloader);
+            if (progress != null)
+            {
+                downloader.ProgressChanged += progress.Report;
+                progress.Report(InitialDownloadProgress.Instance);
+            }
+            var result = downloader.Download(uri, destination);
+            if (result.Status == DownloadStatus.Failed)
+            {
+                throw result.Exception;
+            }
+        }
+
+        private async Task DownloadObjectAsyncImpl(
+            string uri,
+            Stream destination,
+            DownloadObjectOptions options,
+            CancellationToken cancellationToken,
+            IProgress<IDownloadProgress> progress)
+        {
+            destination.CheckNotNull(nameof(destination));
+            var downloader = new MediaDownloader(Service);
+            options?.ModifyDownloader(downloader);
+            if (progress != null)
+            {
+                downloader.ProgressChanged += progress.Report;
+                // Avoid reporting progress synchronously in the original call.
+                await Task.Yield();
+                progress.Report(InitialDownloadProgress.Instance);
+            }
```

Comment:
Threading question, are the progress report callbacks going to happen on the same synchronization context or any?